coderd/inboxnotifications.go

```diff
@@ -0,0 +1,376 @@
+package coderd
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"net/http"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbtime"
+	"github.com/coder/coder/v2/coderd/httpapi"
+	"github.com/coder/coder/v2/coderd/httpmw"
+	"github.com/coder/coder/v2/coderd/pubsub"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/codersdk/wsjson"
+	"github.com/coder/websocket"
+)
+
+// convertInboxNotificationParameters parses and validates the common parameters used in get and list endpoints for inbox notifications
+func convertInboxNotificationParameters(ctx context.Context, logger slog.Logger, targetsParam string, templatesParam string, readStatusParam string) ([]uuid.UUID, []uuid.UUID, string, error) {
+	var targets []uuid.UUID
+	if targetsParam != "" {
+		splitTargets := strings.Split(targetsParam, ",")
+		for _, target := range splitTargets {
+			id, err := uuid.Parse(target)
+			if err != nil {
+				logger.Error(ctx, "unable to parse target id", slog.Error(err))
+				return nil, nil, "", xerrors.New("unable to parse target id")
+			}
+			targets = append(targets, id)
+		}
+	}
+
+	var templates []uuid.UUID
+	if templatesParam != "" {
+		splitTemplates := strings.Split(templatesParam, ",")
+		for _, template := range splitTemplates {
+			id, err := uuid.Parse(template)
+			if err != nil {
+				logger.Error(ctx, "unable to parse template id", slog.Error(err))
+				return nil, nil, "", xerrors.New("unable to parse template id")
+			}
+			templates = append(templates, id)
+		}
+	}
+
+	if readStatusParam != "" {
+		readOptions := []string{
+			string(database.InboxNotificationReadStatusRead),
+			string(database.InboxNotificationReadStatusUnread),
+			string(database.InboxNotificationReadStatusAll),
+		}
+
+		if !slices.Contains(readOptions, readStatusParam) {
+			logger.Error(ctx, "unable to parse read status")
+			return nil, nil, "", xerrors.New("unable to parse read status")
+		}
+	}
+
+	return targets, templates, readStatusParam, nil
+}
+
+// convertInboxNotificationResponse works as a util function to transform a database.InboxNotification to codersdk.InboxNotification
+func convertInboxNotificationResponse(ctx context.Context, logger slog.Logger, notif database.InboxNotification) codersdk.InboxNotification {
+	return codersdk.InboxNotification{
+		ID:         notif.ID,
+		UserID:     notif.UserID,
+		TemplateID: notif.TemplateID,
+		Targets:    notif.Targets,
+		Title:      notif.Title,
+		Content:    notif.Content,
+		Icon:       notif.Icon,
+		Actions: func() []codersdk.InboxNotificationAction {
+			var actionsList []codersdk.InboxNotificationAction
+			err := json.Unmarshal([]byte(notif.Actions), &actionsList)
+			if err != nil {
+				logger.Error(ctx, "unmarshal inbox notification actions", slog.Error(err))
```

Comment:
Just checking: you _want_ to still return the (possibly empty) list of actions if they can't be unmarshaled, right?